larevt/CalibrationDBI/Providers/DBDataset.h

```diff
@@ -0,0 +1,163 @@
+
+#include <string>
+#include <vector>
+#include <variant>
+#include <memory>
+#include "larevt/CalibrationDBI/Interface/CalibrationDBIFwd.h"
+#include "larevt/CalibrationDBI/IOVData/IOVTimeStamp.h"
+
+namespace lariov
+{
+  class DBDataset
+  {
+
+  public:
+
+    // Typedef
+
+    typedef std::variant<long, double, std::unique_ptr<std::string> > value_type;
+
+    // Nested class representing data from one row.
+
+    class DBRow
+    {
+    public:
+
+      // Constructors.
+
+      DBRow() : fData(0) {}
+      DBRow(const value_type* s) : fData(s) {}
+
+      // Accessors.
+
+      bool isValid() const {return fData != 0;}
+      const value_type& getData(size_t col) const {return fData[col];}
+      const std::string& getStringData(size_t col) const {
+	return *std::get<std::unique_ptr<std::string> >(fData[col]);}
+      long getLongData(size_t col) const {return std::get<long>(fData[col]);}
+      double getDoubleData(size_t col) const {return std::get<double>(fData[col]);}
+
+      // Modifiers.
+
+      void clear() {fData = 0; return;}
+
+    private:
+
+      // Data member.
+
+      const value_type* fData;   // Borrowed referenced from enclosing class.
+    };
+
+  // Back to main class.
+
+  public:
+
+    // Constructors, destructor.
+
+    DBDataset();                                   // Default constructor.
+    DBDataset(void* dataset, bool release=false);  // Initializing constructor.
+    ~DBDataset();                                  // Destructor.
+
+    // Simple accessors.
+
+    const IOVTimeStamp& beginTime() const {return fBeginTime;}
+    const IOVTimeStamp& endTime() const {return fEndTime;}
+    size_t nrows() const {return fNRows;}
+    size_t ncols() const {return fNCols;}
+    const std::vector<std::string>& colNames() const {return fColNames;}
+    const std::vector<std::string>& colTypes() const {return fColTypes;}
+    const std::vector<DBChannelID_t>& channels() const {return fChannels;}
+    const std::vector<value_type>& data() const {return fData;}
+
+    // Determine row and column numbers.
+
+    int getRowNumber(DBChannelID_t ch) const;
+    int getColNumber(const std::string& name) const;
+
+    // Access one row.
+
+    DBRow getRow(size_t row) const {return DBRow(&fData[fNCols*row]);}
+
+    // Modifiers.
+
+    void clear();                                    // Return to default-constructed state.
+    void update(void* dataset, bool release=false);  // Reinitialize.
+    void setBeginTime(const IOVTimeStamp& t) {fBeginTime = t;}
+    void setEndTime(const IOVTimeStamp& t) {fEndTime = t;}
+    void setNRows(size_t nrows) {fNRows = nrows;}
+    void setNCols(size_t ncols) {fNCols = ncols;}
+
+    // Modifiable collection accessors.
+
+    std::vector<std::string>& colNames() {return fColNames;}
+    std::vector<std::string>& colTypes() {return fColTypes;}
+    std::vector<DBChannelID_t>& channels() {return fChannels;}
+    std::vector<value_type>& data() {return fData;}
```

Comment:
Suggest removing accessors that allow modification of internal data.  Are they needed?